     Problem Description
    
        Use sklearn.datasets iris flower dataset to train your model using logistic regression. You need to figure out the accuracy of your model and use that to predict different samples in your test dataset. In iris dataset there are 150 samples containing following features,
    1. Sepal Length
    2. Sepal Width
    3. Petal length
    4. Petal width
    
        Using above 4 features you will classify a flower in one of the three categories,
    1. Setosa
    2. Versicolour
    3. Virginica
    Download the iris dataset directly from sklearn.dataset

### Import libraries

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

### Load Iris Dataset

In [2]:
from sklearn.datasets import load_iris

In [3]:
iris = load_iris()
iris

{'data': array([[5.1, 3.5, 1.4, 0.2],
        [4.9, 3. , 1.4, 0.2],
        [4.7, 3.2, 1.3, 0.2],
        [4.6, 3.1, 1.5, 0.2],
        [5. , 3.6, 1.4, 0.2],
        [5.4, 3.9, 1.7, 0.4],
        [4.6, 3.4, 1.4, 0.3],
        [5. , 3.4, 1.5, 0.2],
        [4.4, 2.9, 1.4, 0.2],
        [4.9, 3.1, 1.5, 0.1],
        [5.4, 3.7, 1.5, 0.2],
        [4.8, 3.4, 1.6, 0.2],
        [4.8, 3. , 1.4, 0.1],
        [4.3, 3. , 1.1, 0.1],
        [5.8, 4. , 1.2, 0.2],
        [5.7, 4.4, 1.5, 0.4],
        [5.4, 3.9, 1.3, 0.4],
        [5.1, 3.5, 1.4, 0.3],
        [5.7, 3.8, 1.7, 0.3],
        [5.1, 3.8, 1.5, 0.3],
        [5.4, 3.4, 1.7, 0.2],
        [5.1, 3.7, 1.5, 0.4],
        [4.6, 3.6, 1. , 0.2],
        [5.1, 3.3, 1.7, 0.5],
        [4.8, 3.4, 1.9, 0.2],
        [5. , 3. , 1.6, 0.2],
        [5. , 3.4, 1.6, 0.4],
        [5.2, 3.5, 1.5, 0.2],
        [5.2, 3.4, 1.4, 0.2],
        [4.7, 3.2, 1.6, 0.2],
        [4.8, 3.1, 1.6, 0.2],
        [5.4, 3.4, 1.5, 0.4],
        [5.2, 4.1, 1.5, 0.1],
  

In [4]:
X = iris.data

X

array([[5.1, 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2],
       [4.7, 3.2, 1.3, 0.2],
       [4.6, 3.1, 1.5, 0.2],
       [5. , 3.6, 1.4, 0.2],
       [5.4, 3.9, 1.7, 0.4],
       [4.6, 3.4, 1.4, 0.3],
       [5. , 3.4, 1.5, 0.2],
       [4.4, 2.9, 1.4, 0.2],
       [4.9, 3.1, 1.5, 0.1],
       [5.4, 3.7, 1.5, 0.2],
       [4.8, 3.4, 1.6, 0.2],
       [4.8, 3. , 1.4, 0.1],
       [4.3, 3. , 1.1, 0.1],
       [5.8, 4. , 1.2, 0.2],
       [5.7, 4.4, 1.5, 0.4],
       [5.4, 3.9, 1.3, 0.4],
       [5.1, 3.5, 1.4, 0.3],
       [5.7, 3.8, 1.7, 0.3],
       [5.1, 3.8, 1.5, 0.3],
       [5.4, 3.4, 1.7, 0.2],
       [5.1, 3.7, 1.5, 0.4],
       [4.6, 3.6, 1. , 0.2],
       [5.1, 3.3, 1.7, 0.5],
       [4.8, 3.4, 1.9, 0.2],
       [5. , 3. , 1.6, 0.2],
       [5. , 3.4, 1.6, 0.4],
       [5.2, 3.5, 1.5, 0.2],
       [5.2, 3.4, 1.4, 0.2],
       [4.7, 3.2, 1.6, 0.2],
       [4.8, 3.1, 1.6, 0.2],
       [5.4, 3.4, 1.5, 0.4],
       [5.2, 4.1, 1.5, 0.1],
       [5.5, 4.2, 1.4, 0.2],
       [4.9, 3

In [5]:
y = iris.target

y

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

In [6]:
# Convert the target labels to corresponding class names

class_names = iris.target_names[y]

In [7]:
# Create a DataFrame for better visualization

df = pd.DataFrame(data = np.c_[X, class_names], columns=iris.feature_names + ['species'])

In [8]:
df

sepal length (cm) sepal width (cm) petal length (cm) petal width (cm)   
0                 5.1              3.5               1.4              0.2  \
1                 4.9              3.0               1.4              0.2   
2                 4.7              3.2               1.3              0.2   
3                 4.6              3.1               1.5              0.2   
4                 5.0              3.6               1.4              0.2   
..                ...              ...               ...              ...   
145               6.7              3.0               5.2              2.3   
146               6.3              2.5               5.0              1.9   
147               6.5              3.0               5.2              2.0   
148               6.2              3.4               5.4              2.3   
149               5.9              3.0               5.1              1.8   

       species  
0       setosa  
1       setosa  
2       setosa  
3       setosa  
4       setosa  
..         ...  
145  virginica  
146  virginica  
147  virginica  
148  virginica  
149  virginica  

[150 rows x 5 columns]

In [9]:
df.shape

(150, 5)

In [10]:
iris.feature_names

['sepal length (cm)',
 'sepal width (cm)',
 'petal length (cm)',
 'petal width (cm)']

In [11]:
iris.target_names

array(['setosa', 'versicolor', 'virginica'], dtype='<U10')

### Data Exploration 

In [12]:
df.columns

Index(['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)',
       'petal width (cm)', 'species'],
      dtype='object')

In [13]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   sepal length (cm)  150 non-null    object
 1   sepal width (cm)   150 non-null    object
 2   petal length (cm)  150 non-null    object
 3   petal width (cm)   150 non-null    object
 4   species            150 non-null    object
dtypes: object(5)
memory usage: 6.0+ KB


In [14]:
df.head()

sepal length (cm) sepal width (cm) petal length (cm) petal width (cm)   
0               5.1              3.5               1.4              0.2  \
1               4.9              3.0               1.4              0.2   
2               4.7              3.2               1.3              0.2   
3               4.6              3.1               1.5              0.2   
4               5.0              3.6               1.4              0.2   

  species  
0  setosa  
1  setosa  
2  setosa  
3  setosa  
4  setosa

In [15]:
df.tail()

sepal length (cm) sepal width (cm) petal length (cm) petal width (cm)   
145               6.7              3.0               5.2              2.3  \
146               6.3              2.5               5.0              1.9   
147               6.5              3.0               5.2              2.0   
148               6.2              3.4               5.4              2.3   
149               5.9              3.0               5.1              1.8   

       species  
145  virginica  
146  virginica  
147  virginica  
148  virginica  
149  virginica

In [16]:
# Statistical Analysis

df.describe()

sepal length (cm) sepal width (cm) petal length (cm) petal width (cm)   
count                150              150               150              150  \
unique                35               23                43               22   
top                  5.0              3.0               1.4              0.2   
freq                  10               26                13               29   

       species  
count      150  
unique       3  
top     setosa  
freq        50

### Data Cleaning and Preprocessing

In [17]:
df.isnull()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)   
0                False             False              False             False  \
1                False             False              False             False   
2                False             False              False             False   
3                False             False              False             False   
4                False             False              False             False   
..                 ...               ...                ...               ...   
145              False             False              False             False   
146              False             False              False             False   
147              False             False              False             False   
148              False             False              False             False   
149              False             False              False             False   

     species  
0      False  
1      False  
2      False  
3      False  
4      False  
..       ...  
145    False  
146    False  
147    False  
148    False  
149    False  

[150 rows x 5 columns]

In [18]:
df.isna().sum()

sepal length (cm)    0
sepal width (cm)     0
petal length (cm)    0
petal width (cm)     0
species              0
dtype: int64

### ML Model Using Classification Algorithms 

#### Dataset Splitting 

In [19]:
# import libraries

from sklearn.model_selection import train_test_split

In [20]:
# split the dataset for performing training and testing

X_train, X_test, y_train, y_test = train_test_split(X, y, train_size = 0.25, random_state = 42)

In [21]:
X_train

array([[6.4, 3.2, 5.3, 2.3],
       [5.1, 3.8, 1.6, 0.2],
       [6.9, 3.1, 5.4, 2.1],
       [5.9, 3. , 4.2, 1.5],
       [6.5, 3. , 5.2, 2. ],
       [5.7, 2.6, 3.5, 1. ],
       [5.2, 2.7, 3.9, 1.4],
       [6.1, 3. , 4.6, 1.4],
       [4.5, 2.3, 1.3, 0.3],
       [6.6, 2.9, 4.6, 1.3],
       [5.5, 2.6, 4.4, 1.2],
       [5.3, 3.7, 1.5, 0.2],
       [5.6, 3. , 4.1, 1.3],
       [7.3, 2.9, 6.3, 1.8],
       [6.7, 3.3, 5.7, 2.1],
       [5.1, 3.7, 1.5, 0.4],
       [4.9, 2.4, 3.3, 1. ],
       [6.7, 3.3, 5.7, 2.5],
       [7.2, 3. , 5.8, 1.6],
       [4.9, 3.6, 1.4, 0.1],
       [6.7, 3.1, 5.6, 2.4],
       [4.9, 3. , 1.4, 0.2],
       [6.9, 3.1, 4.9, 1.5],
       [7.4, 2.8, 6.1, 1.9],
       [6.3, 2.9, 5.6, 1.8],
       [5.7, 2.8, 4.1, 1.3],
       [6.5, 3. , 5.5, 1.8],
       [6.3, 2.3, 4.4, 1.3],
       [6.4, 2.9, 4.3, 1.3],
       [5.6, 2.8, 4.9, 2. ],
       [5.9, 3. , 5.1, 1.8],
       [5.4, 3.4, 1.7, 0.2],
       [6.1, 2.8, 4. , 1.3],
       [4.9, 2.5, 4.5, 1.7],
       [5.8, 4

In [22]:
X_test

array([[6.1, 2.8, 4.7, 1.2],
       [5.7, 3.8, 1.7, 0.3],
       [7.7, 2.6, 6.9, 2.3],
       [6. , 2.9, 4.5, 1.5],
       [6.8, 2.8, 4.8, 1.4],
       [5.4, 3.4, 1.5, 0.4],
       [5.6, 2.9, 3.6, 1.3],
       [6.9, 3.1, 5.1, 2.3],
       [6.2, 2.2, 4.5, 1.5],
       [5.8, 2.7, 3.9, 1.2],
       [6.5, 3.2, 5.1, 2. ],
       [4.8, 3. , 1.4, 0.1],
       [5.5, 3.5, 1.3, 0.2],
       [4.9, 3.1, 1.5, 0.1],
       [5.1, 3.8, 1.5, 0.3],
       [6.3, 3.3, 4.7, 1.6],
       [6.5, 3. , 5.8, 2.2],
       [5.6, 2.5, 3.9, 1.1],
       [5.7, 2.8, 4.5, 1.3],
       [6.4, 2.8, 5.6, 2.2],
       [4.7, 3.2, 1.6, 0.2],
       [6.1, 3. , 4.9, 1.8],
       [5. , 3.4, 1.6, 0.4],
       [6.4, 2.8, 5.6, 2.1],
       [7.9, 3.8, 6.4, 2. ],
       [6.7, 3. , 5.2, 2.3],
       [6.7, 2.5, 5.8, 1.8],
       [6.8, 3.2, 5.9, 2.3],
       [4.8, 3. , 1.4, 0.3],
       [4.8, 3.1, 1.6, 0.2],
       [4.6, 3.6, 1. , 0.2],
       [5.7, 4.4, 1.5, 0.4],
       [6.7, 3.1, 4.4, 1.4],
       [4.8, 3.4, 1.6, 0.2],
       [4.4, 3

In [23]:
y_train

array([2, 0, 2, 1, 2, 1, 1, 1, 0, 1, 1, 0, 1, 2, 2, 0, 1, 2, 2, 0, 2, 0,
       1, 2, 2, 1, 2, 1, 1, 2, 2, 0, 1, 2, 0, 1, 2])

In [24]:
y_test

array([1, 0, 2, 1, 1, 0, 1, 2, 1, 1, 2, 0, 0, 0, 0, 1, 2, 1, 1, 2, 0, 2,
       0, 2, 2, 2, 2, 2, 0, 0, 0, 0, 1, 0, 0, 2, 1, 0, 0, 0, 2, 1, 1, 0,
       0, 1, 2, 2, 1, 2, 1, 2, 1, 0, 2, 1, 0, 0, 0, 1, 2, 0, 0, 0, 1, 0,
       1, 2, 0, 1, 2, 0, 2, 2, 1, 1, 2, 1, 0, 1, 2, 0, 0, 1, 1, 0, 2, 0,
       0, 1, 1, 2, 1, 2, 2, 1, 0, 0, 2, 2, 0, 0, 0, 1, 2, 0, 2, 2, 0, 1,
       1, 2, 1])

### 1) Logistic Regression

In [25]:
# import libraries

from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score 

In [26]:
# Create an object 

log_model = LogisticRegression()

log_model

LogisticRegression()

In [27]:
# Train the logistic regression model on the training data.

log_model.fit(X_train, y_train)

LogisticRegression()

In [28]:
# Predict the classes for the test data

y_pred_log = log_model.predict(X_test)

In [29]:
print("The predicted classes of the test data : ")

y_pred_log

The predicted classes of the test data : 


array([1, 0, 2, 1, 1, 0, 1, 2, 1, 1, 2, 0, 0, 0, 0, 1, 2, 1, 1, 2, 0, 2,
       0, 2, 2, 2, 2, 2, 0, 0, 0, 0, 1, 0, 0, 2, 1, 0, 0, 0, 2, 1, 1, 0,
       0, 1, 2, 2, 1, 2, 1, 2, 1, 0, 2, 1, 0, 0, 0, 1, 2, 0, 0, 0, 1, 0,
       1, 2, 0, 1, 2, 0, 2, 2, 1, 1, 2, 1, 0, 1, 2, 0, 0, 1, 2, 0, 2, 0,
       0, 2, 1, 2, 2, 2, 2, 1, 0, 0, 1, 2, 0, 0, 0, 1, 2, 0, 2, 2, 0, 1,
       1, 2, 1])

In [30]:
# Calculate the accuracy of the model

print("The accuracy of the model : ")

log_accuracy_model = accuracy_score(y_test, y_pred_log)

log_accuracy_model * 100

The accuracy of the model : 


96.46017699115043

In [31]:
# Lets create a sample data to predict the category

# Added 3 samples of data 

sample_data_log = np.array([[5.1, 3.5, 1.4, 0.2],
                    [6.2, 2.9, 4.3, 1.3],
                    [7.3, 2.8, 6.3, 1.8]])

In [32]:
sample_data_log

array([[5.1, 3.5, 1.4, 0.2],
       [6.2, 2.9, 4.3, 1.3],
       [7.3, 2.8, 6.3, 1.8]])

##### The array gives the corresponding sepal length (cm)	, sepal width (cm),	petal length (cm), 	petal width (cm) for prediction

In [33]:
# Predict the categories for the data 

sample_predictions_log = log_model.predict(sample_data_log)

sample_predictions_log

array([0, 1, 2])

In [34]:
sample_class_names_log = iris.target_names[sample_predictions_log]

sample_class_names_log

array(['setosa', 'versicolor', 'virginica'], dtype='<U10')

In [35]:
iris.target_names

array(['setosa', 'versicolor', 'virginica'], dtype='<U10')

### 2) Decision Tree

In [36]:
# Import Library

from sklearn.tree import DecisionTreeClassifier

#### Dataset Splitting

In [37]:
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size = 0.55, random_state = 47)

In [38]:
# Create an object for decision tree model

tree_model = DecisionTreeClassifier()

tree_model

DecisionTreeClassifier()

In [39]:
# Fit the model to the training data

tree_model.fit(X_train, y_train)

DecisionTreeClassifier()

In [40]:
# Predict the classes for test data

y_pred = tree_model.predict(X_test)

y_pred

array([0, 0, 1, 0, 0, 2, 1, 2, 1, 0, 2, 0, 1, 2, 2, 2, 1, 0, 0, 0, 2, 1,
       2, 1, 2, 2, 1, 2, 0, 0, 2, 1, 2, 0, 2, 0, 0, 0, 2, 0, 1, 0, 2, 0,
       2, 0, 2, 0, 2, 2, 0, 1, 2, 1, 0, 0, 1, 2, 2, 0, 0, 0, 1, 1, 0, 0,
       2, 2, 1, 0, 1, 2, 2, 2, 0, 1, 0, 1, 0, 2, 2, 2, 1])

In [41]:
# Accuracy of the model

print("Accuracy of the model : ")

accuracy = accuracy_score(y_test, y_pred)

accuracy * 100

Accuracy of the model : 


96.3855421686747

An accuracy of 1.0 means that the model achieved a perfect classification on the test dataset.

In [42]:
# Lets create a sample data to predict the category

tree_sample_data = np.array([[5.1, 3.5, 1.4, 0.2],   
                        [6.2, 2.9, 4.3, 1.3],   
                        [7.3, 2.8, 6.3, 1.8]])

##### The array gives the corresponding sepal length (cm)	, sepal width (cm),	petal length (cm), 	petal width (cm) for prediction

In [43]:
tree_sample_data

array([[5.1, 3.5, 1.4, 0.2],
       [6.2, 2.9, 4.3, 1.3],
       [7.3, 2.8, 6.3, 1.8]])

In [44]:
# Predict the categories for the sample data

tree_sample_prediction = tree_model.predict(tree_sample_data)

In [45]:
tree_sample_prediction

array([0, 1, 2])

In [46]:
tree_sample_class_names = iris.target_names[tree_sample_prediction]

In [47]:
tree_sample_class_names

array(['setosa', 'versicolor', 'virginica'], dtype='<U10')

### 3) Random Forest

In [48]:
# import Library

from sklearn.ensemble import RandomForestClassifier

In [49]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.35, random_state = 45)

In [50]:
# Create a random forest model

rf_model = RandomForestClassifier(n_estimators = 100, random_state = 50)

rf_model

RandomForestClassifier(random_state=50)

In [51]:
# Fit the model to the training data 

rf_model.fit(X_train, y_train)

RandomForestClassifier(random_state=50)

In [52]:
# Predict the classes for the test data

y_pred_rf = rf_model.predict(X_test)

y_pred_rf

array([0, 0, 2, 0, 0, 0, 0, 2, 2, 2, 0, 2, 2, 2, 1, 0, 2, 2, 0, 1, 1, 1,
       2, 1, 0, 2, 1, 1, 0, 1, 1, 1, 1, 2, 0, 2, 0, 0, 1, 0, 0, 1, 0, 1,
       1, 0, 2, 1, 2, 0, 0, 1, 2])

In [53]:
# Accuracy of the model

print("The accuracy of the model : ")

rf_accuracy = accuracy_score(y_test,y_pred_rf)

rf_accuracy * 100

The accuracy of the model : 


96.22641509433963

In [54]:
# Sample data for prediction

rf_sample_data = np.array([[5.1, 3.5, 1.4, 0.2],   
                        [6.2, 2.9, 4.3, 1.3],   
                        [7.3, 2.8, 6.3, 1.8]])

In [55]:
rf_sample_data

array([[5.1, 3.5, 1.4, 0.2],
       [6.2, 2.9, 4.3, 1.3],
       [7.3, 2.8, 6.3, 1.8]])

In [56]:
# Prediction fro the model

prediction_rf = rf_model.predict(rf_sample_data)

prediction_rf

array([0, 1, 2])

In [57]:
# Map the numerical labels to their corresponding class names

rf_class_names = iris.target_names
prediction_rf_names = [rf_class_names[pred] for pred in prediction_rf]

prediction_rf_names

['setosa', 'versicolor', 'virginica']

### 4) Support Vector Machine (SVM)

In [58]:
# import libraries

from sklearn.svm import SVC

#### Dataset Splitting

In [59]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.42, random_state = 45)

In [60]:
# Create an object for support vector machine classifier

svm_model = SVC(random_state = 42)

svm_model 

SVC(random_state=42)

In [61]:
# Train the model on the training data

svm_model.fit(X_train, y_train)

SVC(random_state=42)

In [62]:
# Prediction on the test dataset

svm_y_pred = svm_model.predict(X_test)

svm_y_pred

array([0, 0, 2, 0, 0, 0, 0, 2, 1, 2, 0, 2, 2, 2, 1, 0, 2, 2, 0, 1, 1, 1,
       2, 1, 0, 2, 1, 1, 0, 1, 1, 1, 1, 2, 0, 2, 0, 0, 1, 0, 0, 1, 0, 1,
       1, 0, 2, 1, 2, 0, 0, 1, 2, 0, 2, 0, 0, 1, 0, 1, 1, 1, 1])

In [63]:
# Accuracy of the model

print("The accuracy of the model : ")

accuracy_svm_pred = accuracy_score(y_test, svm_y_pred)

accuracy_svm_pred * 100

The accuracy of the model : 


93.65079365079364

In [64]:
# Use the trained model to make predictions on new samples

svm_new_samples = np.array([[5.1, 3.5, 1.4, 0.2], 
                        [6.3, 3.3, 6.0, 2.5], 
                        [7.2, 3.0, 5.8, 1.6]])

In [65]:
svm_new_samples

array([[5.1, 3.5, 1.4, 0.2],
       [6.3, 3.3, 6. , 2.5],
       [7.2, 3. , 5.8, 1.6]])

In [66]:
# Make predictions on the new samples using the trained model

svm_new_predictions = svm_model.predict(svm_new_samples)

svm_new_predictions

array([0, 2, 2])

In [67]:
# Map the numerical labels to their corresponding class names

svm_class_names = iris.target_names

svm_predicted_classes = [svm_class_names[prd] for prd in svm_new_predictions]

In [68]:
svm_predicted_classes

['setosa', 'virginica', 'virginica']

### 5) K-Nearest Neighbors(KNN)

In [69]:
# Import library

from sklearn.neighbors import KNeighborsClassifier

#### Dataset Splitting

In [70]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.34, random_state = 42)

In [71]:
# # Initialize the KNN classifier with k=3 

knn_model = KNeighborsClassifier(n_neighbors=3)
knn_model

KNeighborsClassifier(n_neighbors=3)

In [72]:
# Train the model on the training data

knn_model.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=3)

In [73]:
# Make predictions on the test set

y_pred_knn = knn_model.predict(X_test)

y_pred_knn

array([1, 0, 2, 1, 1, 0, 1, 2, 1, 1, 2, 0, 0, 0, 0, 1, 2, 1, 1, 2, 0, 2,
       0, 2, 2, 2, 2, 2, 0, 0, 0, 0, 1, 0, 0, 2, 1, 0, 0, 0, 2, 1, 1, 0,
       0, 1, 1, 2, 1, 2, 1, 2])

In [74]:
# Calculate the accuracy of the model

print("The accuracy of the model : ")

knn_accuracy = accuracy_score(y_test, y_pred_knn)

knn_accuracy * 100

The accuracy of the model : 


98.07692307692307

In [75]:
# Use the trained model to make predictions on new samples

knn_new_samples = np.array([[5.1, 3.5, 1.4, 0.2], 
                        [6.3, 3.3, 6.0, 2.5],
                        [7.2, 3.0, 5.8, 1.6]])

In [76]:
knn_new_samples

array([[5.1, 3.5, 1.4, 0.2],
       [6.3, 3.3, 6. , 2.5],
       [7.2, 3. , 5.8, 1.6]])

In [77]:
# Make predictions on the samples data using the trained model

knn_new_predictions = knn_model.predict(knn_new_samples)

knn_new_predictions

array([0, 2, 2])

In [78]:
# Map the numerical labels to their corresponding class names

knn_class_names = iris.target_names

knn_predicted_classes = [knn_class_names[pre] for pre in knn_new_predictions]

knn_predicted_classes

['setosa', 'virginica', 'virginica']

### 6) Naive Bayes algorithm

In [79]:
from sklearn.naive_bayes import GaussianNB

#### Dataset splitting

In [80]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.33, random_state = 42)

In [81]:
# Create an object 

nb_model = GaussianNB()

nb_model

GaussianNB()

In [82]:
# Train the model on the training data

nb_model.fit(X_train, y_train)

GaussianNB()

In [83]:
# Make predictions on the test set

nb_y_pred = nb_model.predict(X_test)

nb_y_pred

array([1, 0, 2, 1, 1, 0, 1, 2, 1, 1, 2, 0, 0, 0, 0, 2, 2, 1, 1, 2, 0, 2,
       0, 2, 2, 2, 2, 2, 0, 0, 0, 0, 1, 0, 0, 2, 1, 0, 0, 0, 2, 1, 1, 0,
       0, 1, 1, 2, 1, 2])

In [84]:
# Calculate the accuracy of the model 

print("The accuracy of the model : ")

accuracy_nb = accuracy_score(y_test, nb_y_pred)

accuracy_nb * 100

The accuracy of the model : 


96.0

In [85]:
# Use the trained model to make predictions on new samples

nb_new_samples = np.array([[5.1, 3.5, 1.4, 0.2], 
                        [6.3, 3.3, 6.0, 2.5],  
                        [7.2, 3.0, 5.8, 1.6]])

In [86]:
nb_new_samples

array([[5.1, 3.5, 1.4, 0.2],
       [6.3, 3.3, 6. , 2.5],
       [7.2, 3. , 5.8, 1.6]])

In [87]:
# Make predictions on the new samples using the trained model

nb_new_predictions = nb_model.predict(nb_new_samples)

nb_new_predictions

array([0, 2, 2])

In [88]:
# Map the numerical labels to their corresponding class names

nb_class_names = iris.target_names

nb_predicted_classes = [nb_class_names[label] for label in nb_new_predictions]

nb_predicted_classes

['setosa', 'virginica', 'virginica']

#### Observations

When using Logistic Regression, the model achieved an accuracy of approximately 96.46%.

The Decision Tree algorithm yielded an accuracy of about 96.39% for the predictions.

Applying the Random Forest algorithm resulted in an accuracy of around 96.23% for the model's predictions.

Utilizing the Support Vector Machine (SVM) algorithm, the model achieved an accuracy of approximately 93.65%.

The K-Nearest Neighbors (KNN) algorithm demonstrated high accuracy, achieving around 98.08% for the model's predictions.

The Naive Bayes algorithm achieved an accuracy of 96.0% for the model's predictions.

